File: app/src/main/java/com/example/newbiechen/nbreader/uilts/StringUtil.kt
package com.example.newbiechen.nbreader.uilts

import android.content.Context
import com.example.newbiechen.nbreader.R
import java.text.DecimalFormat
import kotlin.math.log10
import kotlin.math.pow

/**
 *  author : newbiechen
 *  date : 2019-08-06 15:29
 *  description :
 */

object StringUtil {

    fun size2Str(size: Long): String {
        if (size <= 0) return "0"
        val units = arrayOf("B", "KB", "M", "G", "T")
        //计算单位的，原理是利用lg,公式是 lg(1024^n) = nlg(1024)，最后 nlg(1024)/lg(1024) = n。
        val digitGroups = (log10(size.toDouble()) / log10(1024.0)).toInt()
        //计算原理是，size/单位值。单位值指的是:比如说b = 1024,KB = 1024^2
        return DecimalFormat("#,##0.##").format(
            size / 1024.0.pow(digitGroups.toDouble())
        ) + " " + units[digitGroups]
    }
}

File: app/src/main/cpp/formatplugin/util/StringUtil.cpp
// author : newbiechen
// date : 2019-09-24 16:22
// description : 
//

#include "StringUtil.h"

bool StringUtil::startsWith(const std::string &str, const std::string &start) {
    return
            (start.length() <= str.length()) &&
            #if __GNUC__ == 2
            (str.compare(start, 0, start.length()) == 0);
            #else
            (str.compare(0, start.length(), start) == 0);
            #endif
}

bool StringUtil::endsWith(const std::string &str, const std::string &end) {
    return
            (end.length() <= str.length()) &&
            #if __GNUC__ == 2
            (str.compare(end, str.length() - end.length(), end.length()) == 0);
            #else
            (str.compare(str.length() - end.length(), end.length(), end) == 0);
            #endif
}